packages/frontend/src/redux/actions/staking.js

```diff
@@ -533,3 +524,85 @@ export const handleUpdateCurrent = (accountId) => async (dispatch, getState) =>
     let currentAccount = selectStakingCurrentAccountbyAccountId(getState(), { accountId });
     dispatch(staking.updateCurrent({ currentAccount }));
 };
+
+export const getValidatorFarmData = (validator, accountId) => async (dispatch, getState) => {
+
+    if (validator?.version !== FARMING_VALIDATOR_VERSION || !accountId) return;
+
+    const poolSummary = await validator.contract.get_pool_summary();
+    const farms = await validator.contract.get_farms({ from_index: 0, limit: 300 });
+
+    const list = await Promise.all(farms.map(({ token_id, farm_id, active }) => {
+        try {
+            dispatch(fetchToken({ contractName: token_id }));
+            return validator.contract
+                .get_unclaimed_reward({ account_id: accountId, farm_id })
+                .catch(() => '0')
+                .then((balance) => ({ 
+                    token_id,
+                    balance,
+                    farm_id,
+                    active
+                }));
+        } catch (error) {
+            console.error(error);
+            return ({
+                token_id,
+                farm_id,
+                balance: 0,
+                active
+            });
+        }
+        
+    }));
+
+    const farmData = {
+        poolSummary: {...poolSummary},
+        farmRewards: list,
+    };
+    await dispatch(staking.setValidatorFarmData(validator.accountId, farmData));
+    return farmData;
```

Comment:
Since this puts data into state where it should be selected using a selector, it would be an anti-pattern for the caller to also wait for the returned result.  🔪  this return so that the method doesn't encourage someone to `await` this thunk rather than using selectors to load the data that this thunk populates.